src/rapids_pre_commit_hooks/copyright.py

```diff
@@ -0,0 +1,437 @@
+                if (match := BRANCH_RE.search(branch.name))
+            ),
+            key=lambda i: i[1],
+        )[0]
+    except ValueError:
+        pass
+
+    # Appropriate branch not found
+    warnings.warn(
+        "Could not determine target branch. Try setting the TARGET_BRANCH or "
+        "RAPIDS_BASE_BRANCH environment variable, or setting the rapidsai.baseBranch "
+        "configuration option.",
+        NoTargetBranchWarning,
+    )
+    return None
+
+
+def get_target_branch_upstream_commit(repo, target_branch_arg=None):
+    target_branch = get_target_branch(repo, target_branch_arg)
+    if target_branch is None:
+        try:
+            return repo.head.commit
+        except ValueError:
+            return None
+
+    target_branch_upstream = target_branch.tracking_branch()
+    if target_branch_upstream:
+        return target_branch_upstream.commit
+
+    def try_get_ref(remote):
+        try:
+            return remote.refs[target_branch.name]
+        except IndexError:
+            return None
+
+    try:
+        return max(
+            (upstream for remote in repo.remotes if (upstream := try_get_ref(remote))),
+            key=lambda upstream: upstream.commit.committed_datetime,
+        ).commit
+    except ValueError:
+        pass
+
+    return target_branch.commit
+
+
+def get_changed_files(target_branch_arg):
+    try:
+        repo = git.Repo()
+    except git.InvalidGitRepositoryError:
+        return {
+            os.path.relpath(os.path.join(dirpath, filename), "."): None
+            for dirpath, dirnames, filenames in os.walk(".")
+            for filename in filenames
+        }
+
+    changed_files = {f: None for f in repo.untracked_files}
+    target_branch_upstream_commit = get_target_branch_upstream_commit(
+        repo, target_branch_arg
+    )
+    if target_branch_upstream_commit is None:
+        changed_files.update({blob.path: None for _, blob in repo.index.iter_blobs()})
+        return changed_files
+
+    diffs = target_branch_upstream_commit.diff(
+        other=None,
+        merge_base=True,
+        find_copies=True,
+        find_copies_harder=True,
+        find_renames=True,
+    )
+    for diff in diffs:
+        if diff.change_type == "A":
+            changed_files[diff.b_path] = None
+        elif diff.change_type != "D":
+            changed_files[diff.b_path] = diff.a_blob
+
+    return changed_files
+
+
+def find_blob(tree, filename):
+    d1, d2 = os.path.split(filename)
+    split = [d2]
+    while d1:
+        d1, d2 = os.path.split(d1)
+        split.insert(0, d2)
+
+    while len(split) > 1:
+        component = split.pop(0)
+        try:
+            tree = next(t for t in tree.trees if t.name == component)
+        except StopIteration:
+            return None
+
+    try:
+        return next(blob for blob in tree.blobs if blob.name == split[0])
+    except StopIteration:
+        return None
+
+
+def get_file_last_modified(commit, filename):
```

Comment:
I expected this to be a lot simpler. I was thinking you'd get the last modified year like this:

```python
latest_commit = next(repo.iter_commits(paths=linter.filename))
year = latest_commit.committed_datetime.year
```

Are there cases that require us to walk the whole commit history, or can we just look up the latest commit for each file as above? It's totally possible that I'm missing something here. If it comes down to "commit changes in a PR in 2023, revert change in a PR in 2024", I expect this script to use year 2024, so I'm not sure what the parent commits are used for.

Reply:
> Are there cases that require us to walk the whole commit history, or can we just look up the latest commit for each file as above?

This doesn't walk the *entire* history, only until it finds a change in the file that warrants a copyright update. This allows us to pierce through copyright-header-only changes (updating the copyright year when nothing else changed) and even revert them, as this mode does.

If we don't want to do this, then yes, we can simply find the last commit where the file changed.

Reply:
Thanks for the explanation. I think I would prefer the simpler behavior here. You mentioned the runtime was 18 minutes, which is a bit concerning. I would favor simpler code and faster runtime, even if it costs us a bit of functionality (reverting years from copyright-only changes).

Reply:
I just tried to use `iter_commits()` to find when a file was last modified. Unfortunately, it doesn't look like there's any way to force this to skip over `chmod`-only changes, which we agree should not warrant a copyright update. We will still have to do some intelligent checking, and not just blindly take the first commit that `iter_commits()` gives us.

If we have to intelligently look at each commit from `iter_commits()` anyway, we could combine `iter_commits()` with the smart copyright checking that I have now. Perhaps that would be faster than iterating through every single commit as I am now.

Reply:
I'm not too worried about `chmod`-only changes. If we can speed up the copyright batch mode by requiring new years for permissions changes, that is fine with me.

Overall I think I'll accept whatever you think is the right choice here, but I'd prioritize performance. So long as it only generates false positives (copyright years being "too new" for the contents) and not false negatives (not including later years with content changes), we'll be alright.

Reply:
One thing to be aware of is that allowing false positives will cause `--batch` mode to give a different answer than if `pre-commit` mode had been used all along. This is probably fine. I'll implement the simplified logic.